File: templating/tag_contributions/src/main/java/com/vmware/jsonteng/ext/tags/net/Ipv4HostNetmaskTag.java
package com.vmware.jsonteng.ext.tags.net;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;

import com.vmware.jsonteng.ElementResolver;
import com.vmware.jsonteng.TagResolver;
import com.vmware.jsonteng.TemplateEngineException;
import com.vmware.jsonteng.tags.TagBase;

public class Ipv4HostNetmaskTag extends TagBase {
    public static final String name = "ipv4-host-netmask";

    private final ElementResolver elementResolver;

    public Ipv4HostNetmaskTag(TagResolver tagResolver) {
        super(tagResolver);
        elementResolver = tagResolver.getElementResolver();
    }

    @Override
    public Object process(List<?> tagTokens, List<Map<String, ?>> bindingDataList) throws TemplateEngineException {
        int tokenCount = tagTokens.size();
        if (tokenCount < 1) {
            throw new TemplateEngineException(
                    String.format("Tag \"%s\" requires 1 parameter. Parameters given %s", name, tagTokens));
        }
        String resolvedNetwork = (String) this.elementResolver.resolve(
                tagTokens.get(0), bindingDataList);
        String[] parts = resolvedNetwork.split("/");
        int prefix = Integer.parseInt(parts[1]);
        ByteBuffer fourBytes = ByteBuffer.wrap(new byte[4]);
        int netmaskInt = 0xFFFFFFFF << (32 - prefix);
        fourBytes.putInt(netmaskInt);
        Inet4Address netmask;
        try {
            netmask = (Inet4Address) InetAddress.getByAddress(fourBytes.array());
        } catch (UnknownHostException e) {
            throw new TemplateEngineException("Unable to get netmask " + fourBytes);
        }
        return netmask.getHostAddress();
    }
}

File: templating/tag_contributions/src/main/python/jsonteng_contribs/tags/ipv4_host_netmask_tag.py
# Copyright 2019 VMware, Inc.
# SPDX-License-Indentifier: Apache-2.0

import ipaddress

from jsonteng.exception import TemplateEngineException
from jsonteng.tags.tag_base import TagBase


class Ipv4HostNetmaskTag(TagBase):
    """
    Network tag for returning network address, netmask and gateway.
    Gateway is assumed to be the lowest network address.
    """
    name = "ipv4-host-netmask"

    def __init__(self, tag_resolver):
        """
        Construct this tag.
        :param tag_resolver: Tag resolver
        :type tag_resolver: 'TagResolver'
        """
        super().__init__(tag_resolver)
        self._element_resolver = tag_resolver.get_element_resolver()

    def process(self, tag_tokens, binding_data_list):
        """
        Process this tag.
        :param tag_tokens: Tag arguments.
        :type tag_tokens: 'list'
        :param binding_data_list: Binding data used during the processing.
        :type binding_data_list: 'list'
        :return: JSON object
        :rtype: JSON object
        """
        token_count = len(tag_tokens)
        if token_count < 1:
            raise TemplateEngineException(
                "Tag \"{}\" requires 1 parameter."
                " Parameters given {}".format(Ipv4HostNetmaskTag.name, tag_tokens))
        network = ipaddress.ip_network(self._element_resolver.resolve(
            tag_tokens[0], binding_data_list))
        return str(network.netmask)
